torch/_guards.py

```diff
@@ -517,20 +523,37 @@ def clear_frame():
         with unittest.mock.patch.object(
             tc, "frame_summary_stack", []
         ), unittest.mock.patch.object(tc, "loc_in_frame", None):
-            yield
+            try:
+                yield
+            except Exception as e:
+                # Prevent real_stack from getting attached
+                if not hasattr(e, "real_stack"):
+                    e.real_stack = None  # type: ignore[attr-defined]
+                raise
 
     @staticmethod
     @contextlib.contextmanager
     def current_frame(frame_summary):
+        # frame_summary can be None to solely take advantage of real_stack
+        # attachment to thrown exceptions
         tc = TracingContext.get()
         assert (
             tc is not None
         ), "Frame context manager must be called within an ongoing trace."
-        tc.frame_summary_stack.append(frame_summary)
+        if frame_summary is not None:
+            tc.frame_summary_stack.append(frame_summary)
+        old = tc.loc_in_frame
+        tc.loc_in_frame = None
         try:
             yield
+        except Exception as e:
+            if not hasattr(e, "real_stack"):
+                e.real_stack = tc.extract_stack()  # type: ignore[attr-defined]
```

Comment:
Attach the stack if there isn't one already. This ensures we get localized error messages for ALL exceptions, not just Unsupported.